clippy_lints/src/needless_as_bytes.rs

```diff
@@ -0,0 +1,58 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::sugg::Sugg;
+use clippy_utils::ty::is_type_lang_item;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, LangItem};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// It detects useless calls to `str::as_bytes()` before calling `len()` or `is_empty()`.
+    ///
+    /// ### Why is this bad?
+    /// The `len()` and `is_empty()` methods are also directly available on strings.
```

Comment:
I would probably clarify that the `len()` and `is_empty()` methods work exactly the same - because some programmers may assume that `str::len()` returns the number of characters, whereas it actually returns the number of bytes.

Reply:
Yes, I've clarified this and made it totally non-ambiguous:

```rust
    /// ### Why is this bad?
    /// The `len()` and `is_empty()` methods are also directly available on strings, and they
    /// return identical results. In particular, `len()` on a string returns the number of
    /// bytes.
 ```